### PR TITLE
fix(python): keep a space's grid provider alive (#378)

### LINKED ISSUE #378: Segfault in estimate_combined_inverse_trace_inequality_constant on a 3d ALUGrid cube
Uncovered by #376, which makes LAPACKE a hard requirement. This is a pre-existing crash, not a regression from that PR — it was simply unreachable in CI until now.

## Symptom

`gdt_test_python` segfaults (the whole pytest process dies, exit code 8) in
`python/gdt/test/test_hypothesis_grid_quality.py::test_inverse_inequality_constants_are_positive_and_finite`:

```
Fatal Python error: Segmentation fault

Current thread 0x0000739387d80780 (most recent call first):
  File ".../python/gdt/test/test_hypothesis_grid_quality.py", line 147 in test_inverse_inequality_constants_are_positive_and_finite
  File ".../hypothesis/core.py", line 1004 in test
  ...
```

Line 147 is `estimate_combined_inverse_trace_inequality_constant(space)`. Line 146,
`estimate_inverse_inequality_constant(space)`, returns normally for the same space — so the crash is specific to
the *combined inverse trace* estimator, not to the generalized eigen-solver in general.

## Why it only surfaces now

`test_inverse_inequality_constants_are_positive_and_finite` is guarded by `@_needs_eigen_solver`, which probes for
LAPACKE and skips the whole test when it is absent. CI had `HAVE_LAPACKE == 0`, so this test had **never executed**.
The C++ counterpart `dune/gdt/test/misc/tools_grid_quality_estimates.cc` only pins a 2d YASP cube at order 1 and is
green, so the estimator has effectively only ever been exercised on that one configuration.

## Failing configuration

The last grid constructed before the fault was a 3d ALUGrid cube:

```
Created serial ALUGrid<3,3,cube> from macro grid file 'Cartesian ALUGrid via YaspGrid'.
```

It was the third such grid drawn in that test, so the crashing example is a 3d cube ALUGrid with a
`ContinuousLagrangeSpace` of order 1 or 2 (the test draws `order=st.integers(1, 2)`). The exact bounding box and
element counts are not recoverable from the log — a segfault kills the process before hypothesis can print the
falsifying example.

## Where to look

`estimate_combined_inverse_trace_inequality_constant` in `dune/gdt/tools/grid-quality-estimates.hh` (lines 78-115).
It differs from the (working) `estimate_inverse_inequality_constant` in that it assembles a face product matrix by
looping over the intersections of each element:

```cpp
XT::LA::CommonDenseMatrix<double> L2_face_product_matrix(basis->size(), basis->size(), 0.);
DynamicMatrix<double> tmp_L2_face_product_matrix(basis->size(), basis->size(), 0.);
for (auto&& intersection : intersections(space.grid_view(), element)) {
  LocalIntersectionIntegralBilinearForm<I, r>(LocalProductIntegrand<I, r>(1.))
      .apply2(intersection, *basis, *basis, tmp_L2_face_product_matrix);
  ...
}
```

Note also that the face product matrix is only positive *semi*-definite (an order-2 Lagrange basis on a cube has an
interior node whose basis function vanishes on the element boundary), which is why the eigenvalue loop filters
`std::abs(ev) > 1e-7`. That is fine for `dsygv` — it requires only the right hand side to be positive definite — but
it is worth confirming that the semi-definite lhs is not part of the story.

## Reproducing

Needs a build with LAPACKE, ALUGrid and the Python bindings:

```
pytest python/gdt/test/test_hypothesis_grid_quality.py::test_inverse_inequality_constants_are_positive_and_finite
```

A C++ reproducer extending `dune/gdt/test/misc/tools_grid_quality_estimates.cc` to a 3d ALU cube grid at orders 1
and 2 would be the more useful regression test, since it can be run under a debugger and ASAN directly.

## Acceptance criteria

- [ ] `estimate_combined_inverse_trace_inequality_constant` returns a finite, positive constant on a 3d ALUGrid cube for orders 1 and 2 (or throws a proper exception if some configuration is genuinely unsupported).
- [ ] A C++ test pins the estimator on more than the single 2d YASP square it covers today.
- [ ] `test_inverse_inequality_constants_are_positive_and_finite` runs unquarantined over the full `grid_specs` product.

## Evidence

- Run: https://github.com/dune-gdt/dune-gdt/actions/runs/30158429526/job/89679653767 (`Build on Linux (clang22-debug)`, `Test with CTest`)
- 601 of 602 ctest tests pass; `gdt_test_python` is the only failure. The whole C++ suite, including the newly enabled generalized-eigensolver tests, is green.

### PR DESCRIPTION
**Pull request by renefritze**
Fixes #378.

## Root cause

`estimate_combined_inverse_trace_inequality_constant` is not at fault. The crash is a dangling grid on the python side.

Every space binding hands the C++ constructor `grid_provider.leaf_view()`, and a `GridView` is only a handle into the grid the `GridProvider` owns (`std::shared_ptr<GridType> grid_ptr_`). Nothing tied the two lifetimes together, so the line the traceback points at,

```python
space = ContinuousLagrangeSpace(spec.make_grid(), order=order)
```

drops the last reference to the provider at the end of the statement, frees the grid, and leaves the space pointing into it. Every later use is a use-after-free. On YaspGrid the freed memory happens to still read back plausibly — which is why `test_hypothesis_spaces.py:76` has been writing the same thing without incident — and on a 3d ALUGrid cube it does not.

Reproduced under ASAN with a C++ transcription of that python statement (grid provider destroyed, space kept):

```
ERROR: AddressSanitizer: heap-use-after-free on address 0x515000002d90
    #0 Dune::ALU3dGrid<3, 3, hexa, ALUGridNoComm>::leafend<0, All_Partition>() const
    ...
    #5 Dune::GDT::estimate_inverse_inequality_constant<...>(...) dune/gdt/tools/grid-quality-estimates.hh:49
freed by thread T0 here:
    #8 Dune::XT::Grid::GridProvider<Dune::ALUGrid<3, 3, cube, nonconforming, ALUGridNoComm>>::~GridProvider()
```

Note it lands in the *first* estimator there. Which of the two calls falls over is down to allocator reuse, not to the estimator — which is also why `estimate_inverse_inequality_constant` returning normally in CI was misleading.

The estimator itself is fine on the failing configuration: over 550 randomised geometries mirroring what `grid_specs()` draws (bounding boxes with per-axis extents in `[0.1, 100]`, corners in `[-100, 100]`, 1–3 elements per axis; 3d ALU cube, 2d/3d YASP, 2d/3d ALU simplex, orders 1 and 2) both constants came back positive and finite with the grid alive.

## Changes

**Bindings** — `py::keep_alive` between every space factory and its grid provider, on both the `py::init` and the module-level convenience overloads:
`ContinuousLagrangeSpace`, `DiscontinuousLagrangeSpace`, `FiniteVolumeSpace`, `FiniteVolumeSkeletonSpace`, `RaviartThomasSpace`.

**Estimators** (`dune/gdt/tools/grid-quality-estimates.hh`) — both silently returned `inf` (`h * max double`) for an element whose eigenvalues all fell below the `1e-7` cutoff, since `min_ev` was left at `std::numeric_limits<double>::max()`. They now throw a new `Exceptions::tools_error` naming the element instead, per the issue's "or throws a proper exception" clause. Nothing in the sweep above reaches it.

**Tests**

- `dune/gdt/test/misc/tools_grid_quality_estimates.cc` covered a single 2d YASP square at order 1. It now covers 2d and 3d YASP cubes, the 3d ALU cube and 2d/3d ALU simplices at orders 1 and 2, and pins:
  - the closed forms on cube grids — `2√6` / `4√2` (2d order 1), `2√6` / `6√2` (2d order 2), `6` / `6√3` (3d, both orders);
  - that both constants are independent of the mesh width of a uniform partition (they are ratios in which `h` cancels), for cubes and simplices alike;
  - that the 3d ALU cube and the 3d YASP cube over the same box agree — both estimates are purely geometric, so the grid manager must not matter.
- `test_a_space_keeps_its_grid_alive` (`test_hypothesis_spaces.py`) pins the binding contract directly, once per compiled grid implementation, building each space from a temporary provider and forcing a `gc.collect()` before touching it. Run in a subprocess, following the convention already used for `test_cg_interpolation_points_match_dofs_and_lie_in_domain`, so a regression fails that test instead of killing the pytest process.
- `test_inverse_inequality_constants_are_positive_and_finite` still passes its grid provider as a temporary on purpose (binding it to a local would paper over exactly this bug); a comment now says so. It stays unquarantined over the full `grid_specs` product.

## Verification

Built dune-common/geometry/grid/istl/localfunctions/alugrid at the refs the vcpkg overlays pin, against system LAPACKE, and ran the new C++ suite — 7/7 pass, including `inverse_inequality_constants_on_3d_alu_cube_grid`. The five touched binding translation units compile clean. `clang-format` and `ruff` clean.

The python suite itself is not runnable in this environment (no wheel build), so `test_a_space_keeps_its_grid_alive` is verified by construction and by the equivalent C++/ASAN reproducer rather than by a local run — CI is the first place it actually executes.

## Note for reviewers

The same missing-`keep_alive` pattern exists in other bindings that take a `GridProvider` and hand back something holding a grid view (the `xt` grid walkers/functors/filters and several `xt.functions` factories, among others). Those are out of scope here; this PR fixes the space factories, which is what #378 hits.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01SW6ihHXRUL9Wv9q9wux24Y

---
_Generated by [Claude Code](https://claude.ai/code/session_01SW6ihHXRUL9Wv9q9wux24Y)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed Python space and finite-volume bindings to prevent use-after-free when the underlying grid provider is temporary.
  * Improved grid-quality estimate failures with clearer error reporting when eigenvalue solving finds no valid positive eigenvalues.
* **Enhancements**
  * Expanded grid-quality estimate validation with mesh-width and scale invariance checks across cube and simplicial grids, including cross-grid consistency where available.
* **Tests**
  * Added a subprocess regression test to ensure spaces remain usable after provider garbage collection.
  * Removed the skip for the Hypothesis 3D ALUGrid cube-grid quality property test.

### CODE REVIEW

File: dune/gdt/tools/grid-quality-estimates.hh [hunk 1]
```diff
@@ -27,13 +27,56 @@
...
+  for (auto&& ev : evs)
+    if (std::abs(ev) > 1e-7) // TODO: find a better tolerance here!
+      min_ev = std::min(min_ev, ev);
+  // Without this the callers would silently hand out h * (max double) = inf for an element whose eigenvalues all fall
+  // below the tolerance above.
+  DUNE_THROW_IF(min_ev == std::numeric_limits<double>::max(),
+                Exceptions::tools_error,
+                "No eigenvalue of the " << matrix_pair << " matrix pair exceeds the 1e-7 cutoff on the element at "
+                                        << element.geometry().center() << " (diameter " << XT::Grid::diameter(element)
+                                        << "), so no constant can be estimated!");
```
[the remark this answers is not in the document]
Comment (renefritze) [reply dune/gdt/tools/grid-quality-estimates.hh:91]: Confirmed both failure modes against the real code, and the second one is worse than described. Fixed in eac5664.

I dumped the raw H1/L2 spectrum on `[0, s]^3` at order 1/2 (2x2x2, YaspGrid):

| extent | largest &#124;λ&#124; | most negative λ | negative λ passing `abs(ev) > 1e-7` | old `inverse` |
|---|---|---|---|---|
| 1e0 | 1.44e+2 | 0 | 0 | 6.0 ✅ |
| 1e4 | 1.44e-6 | -6.6e-23 | 0 | 6.0 ✅ |
| 1e5 | 1.44e-8 | 0 | 0 | **throws `tools_error`** |
| 1e6 | 1.44e-10 | 0 | 0 | **throws `tools_error`** |
| 1e-3 (order 2) | 7.2e+8 | — | — | **2.2251e-308** |
| 1e-4 (order 2) | 7.2e+10 | **-7.63e-06** | **1** | — |

So the large-element case is exactly as you predicted — at extent 1e5 the whole spectrum is below 1e-7 and nothing survives the filter.

The small-element case doesn't surface as a `NaN` return, though, which is what makes it nastier than the comment suggests: `sqrt` of the negative mode *is* NaN, but the caller does `result = std::max(result, h * std::sqrt(min_ev))`, and `std::max` with a NaN second argument returns the first. The NaN is silently dropped and `result` stays at its `std::numeric_limits<double>::min()` seed — so on `[0, 1e-3]^3` the estimator returned **2.2251e-308** instead of 6. Positive, finite, and completely wrong, which means it sails straight past the `math.isfinite(...) and ... > 0.0` property in `test_hypothesis_grid_quality.py`.

Fix is what you suggested: cutoff at `1e-7 * max|λ|` (above the O(eps·max) noise floor, below the smallest genuine mode, which is `max|λ|` over a mesh-independent constant of order 10–100) and filter on `ev > cutoff` rather than `abs(ev) > cutoff`, so a negative mode can never be selected. The `tools_error` guard stays for the genuinely degenerate case.

On the regression case: rather than one large element I added `check_scale_invariance`, which asserts both constants are unchanged on `[0, s]^d` for s from 1e-5 to 1e6 — these constants are dimensionless, so scale invariance is the property that was actually being violated at both ends. It runs for 2d/3d YaspGrid and the 3d ALUGrid cube at orders 1 and 2. Independently, a 1000-configuration sweep over the geometries `grid_specs()` draws, repeated at five domain scales from 1e-6 to 1e6, is clean with the fix and **fails 444 of those 1000 with the old filter**. Pinned closed forms on the unit box are unchanged.

Worth noting this is pre-existing — the `// TODO: find a better tolerance here!` it replaces predates this branch — but it lives in the helper this PR introduced, so it belongs here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
## `Build docs` fails here and passes on main — investigated, not yet explained

Flagging this before the armed auto-merge can act on a red gate. The bot summary rates the doc-build failure "Related to change: unclear"; I could not reduce that to "unrelated", so here is what I actually established.

### The failure

`Build docs` dies with `nbclient.exceptions.DeadKernelError: Kernel died` while executing `docs/source/example__ESV2007_estimates.md`, aborting sphinx with exit 2. Because it is a dead kernel rather than a cell error, `nb_execution_raise_on_error=False` never gets to write `reports/*.err.log`, so there is **no traceback and no indication of which cell**.

Reproduced twice on `cd6ec27` (jobs 89748257383, 89748857979), both at the same notebook, immediately after `dune_gdt_tutorial_on_data_functions_and_interpolation`. It is deterministic, not flaky.

### What argues it is not this PR

* **The notebook runs clean locally against the exact wheel from the failing run.** I downloaded artifact `wheels_3.13` (built 02:46 by job 89739219065, the same wheel the failing docs job installed), installed it into a 3.13 venv with the `[docs,visualisation]` extras, and executed `example__ESV2007_estimates.md` through `nbclient` in a real Jupyter kernel — all 15 cells, including the five `visualize_function` calls, exit 0.
* **Not memory.** Peak RSS 0.82 GB; the grid is 128 simplices. This job has a documented OOM history (hence `sphinx -j 1`), but this notebook is nowhere near the 16 GB box.
* **Nothing in the diff is on that notebook's execution path.** `estimate_combined_inverse_trace_inequality_constant` and `estimate_element_to_intersection_equivalence_constant` are *imported but never called* there (lines 84-85; the only call site in `docs/` is `example__ipdg_heat_equation.md`, which sorts after ESV2007 and never ran). So the new `Exceptions::tools_error` and the relative eigenvalue cutoff cannot be involved. The three spaces it builds — `DiscontinuousLagrangeSpace`, `RaviartThomasSpace`, `FiniteVolumeSpace` — all take a **named** `grid` variable, so `py::keep_alive` changes no lifetime; it only adds a reference to an object that was already alive.

### What argues it is

* `main` at **629faee2 — this PR's exact base** — ran the same docs job green (run 30170704330, job 89724981832). main's overall red was only `Publish to Test PyPI`, which is not in the gate.
* I checked and discarded both change-independent explanations I could think of: no `uv.lock`, but no relevant dependency (`vtk`, `k3d`, `nbclient`, `myst-nb`, `jupyter-cache`) was published between main's docs run and this one — latest releases are months old. And every DUNE dep is pinned to an exact commit in `.vcpkg-overlays/ports/*/portfile.cmake`, with `dune-xt` in-repo. So this wheel is main's wheel plus the 12 files in this PR.

### Where that leaves it

Deterministic in CI, green on main, identical environment and dependencies — but not reproducible outside CI with the very wheel that fails inside it. The one ingredient I have not reproduced is the CI-only wrapper: `xvfb-run` plus the sphinx/`jupyter-cache` layer around the kernel.

Worth noting that ESV2007 is the first notebook in `reading sources` order to construct `RaviartThomasSpace` and `FiniteVolumeSpace`, so it is the first to exercise those two factories — though the preceding CG tutorial exercises `ContinuousLagrangeSpace` through the same changed code path and passed.

I am not asking for this to be waved through, and I would rather not have it merge red. Useful next steps, in order of cost:

1. Preserve the kernel's stderr in that job (or run it under `faulthandler`) so a dead kernel leaves a signal/stack instead of nothing — the current setup makes this class of failure undiagnosable from the logs alone.
2. Failing that, drop the `keep_alive` commit on a scratch branch and run docs, which isolates the bindings from the rest of the PR in one CI round.

Happy to do either; say which you prefer. @renefritze — flagging because auto-merge is armed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
## The docs kernel death is now confirmed deterministic — pushed diagnostics (option 1 from my earlier comment)

Since [my last comment](https://github.com/dune-gdt/dune-gdt/pull/379#issuecomment-5081750163), `Build docs` has been retried three more times: it reproduced the identical `DeadKernelError: Kernel died` on attempts 3 and 5, and hit an unrelated, transient artifact-lookup 404 on attempt 4 (the docs job never reached notebook execution that time, so it's not evidence either way). That's **3 identical crashes out of 4 runs that got that far** — no reasonable reading of that is "flaky."

With no maintainer reply yet and the evidence now unambiguous, I went ahead with the lower-cost remedy I proposed (option 1), since it only touches CI diagnostics, not the code under test:

- `PYTHONFAULTHANDLER=1` on the `Build docs` step, so a kernel dying from a trappable signal (segfault, abort) now dumps its C-level Python stack to the job log instead of vanishing into a bare `DeadKernelError`.
- A `dmesg`-on-failure step, since `faulthandler` can't catch `SIGKILL` — if the OOM-killer is involved (this job has hit that before), `dmesg` is the only place it's recorded.

Pushed as `74f7b55`. Next failure on this branch should carry either a real traceback or a kernel-log line pointing at the actual cause, instead of nothing. I still haven't touched option 2 (isolating the `keep_alive` commit on a scratch branch) since that needs a branch beyond the one I'm authorized to push to.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
## The diagnostics ran — and found nothing, which is itself informative

Head `74f7b55`'s `Build docs` job failed a 4th time on `example__ESV2007_estimates.md` (job [89790546899](https://github.com/dune-gdt/dune-gdt/actions/runs/30197145256/job/89790546899)), same notebook, same point, identical `DeadKernelError`. Both diagnostics ran and both came back empty:

- **No `faulthandler` output.** I searched the full job log for `Fatal Python error` / `Current thread` (`faulthandler`'s signature output) — absent. `PYTHONFAULTHANDLER=1` traps `SIGSEGV`/`SIGABRT`/`SIGFPE`/`SIGBUS`/`SIGILL`, so this rules those out: the kernel process was not killed by a trappable fault.
- **No `dmesg` entry.** The dump's last 200 lines are entirely VM boot messages from 09:49-09:51; nothing at all is logged between then and the 11:56 crash. No OOM-killer line, no segfault line — the kernel-level ring buffer recorded nothing around the crash.

That combination narrows it further than before: not a segfault/abort, and not the host-level OOM-killer this job has hit previously. What's left is `SIGKILL` from something other than the standard OOM-killer, or — a candidate I hadn't considered until now — a **plain, non-signal process exit**. VTK's default X11 error handler calls `exit()` (not a signal, not a Python exception) on certain X protocol errors, and this notebook is the first one to call `visualize_function` repeatedly (5 times) under `xvfb-run`. That would produce exactly what's observed: sudden death, nothing for `faulthandler` to trap, nothing in `dmesg` since it's a userspace X11/VTK event, not a kernel one. I have no direct evidence for this yet, only that it fits every constraint the diagnostics have established.

I've now exhausted what I can do unilaterally within my own branch — reproducing the exact CI wrapper (`xvfb-run` + this specific `k3d`/VTK headless rendering path under repeated calls) isn't something I can fully replicate outside CI. The two remaining options from my first comment stand:

1. Isolate the `keep_alive` commit on a scratch branch to run docs against just the bindings (needs a branch beyond `claude/issue-378-4dv9ge`, which needs explicit permission).
2. Alternatively, if the VTK/X11 theory is worth testing, capture Xvfb's own log (`xvfb-run --auto-servernum --error-file=...` or similar) in a future run — I can add that myself if wanted, same low-risk diagnostics-only change as this one.

@renefritze — this has now failed identically 4/5 times with two rounds of instrumentation added and no lead beyond ruling things out. I don't want to keep pushing diagnostics into the dark without direction; let me know which of the above (or something else) you'd like.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
Opened both as separate PRs off `main`, per your go-ahead:

- #402 — Xvfb error-log capture (also serves as an uncoupled test: does `Build docs` crash on plain `main`, with zero code from this PR?)
- #403 — isolates just the seven `py::keep_alive` binding files from this PR, dropping the estimator/exceptions/test changes entirely

Watching both; will report back once their `Build docs` runs land.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
## Both diagnostic PRs came back green — `Build docs` did not reproduce on either

- **#402** (plain `main` + diagnostics only, **zero code from this PR**): `Build docs` passed — 22:51:49 to 23:21:34, ~30 min.
- **#403** (`main` + *only* the 7 `py::keep_alive` binding files from this PR, none of the estimator/exceptions/test changes): `Build docs` passed — 22:50:29 to 23:19:54, ~29 min.

Neither branch hit `example__ESV2007_estimates.md`'s `DeadKernelError`, not even once. That's the opposite of what I expected going in — I thought #402 in particular (identical environment, identical notebook, zero diff) had a good chance of reproducing it if the cause really were the untested VTK/k3d/Xvfb path from #393. It didn't.

Putting this together with the 4/5 failure rate on this PR's own branch (`cd6ec27`/`74f7b55`), a few readings are possible:

1. **Something in this PR's full diff is responsible after all** — despite the estimator/exceptions/test changes not being on that notebook's call path, some indirect effect (e.g. a shift in the compiled `_gdt` extension's memory layout from the added template instantiations) could be stressing an already-fragile VTK/X11 path differently. This is speculative and I have no direct evidence for it.
2. **`main` has since picked up an unrelated fix** that happens to also cover this — both diagnostic branches are based on current `main` (`f77b6bc`), which is several commits ahead of this PR's base (`629faee`). If so, merging `main` into this branch and re-running should turn `Build docs` green here too.
3. **It was environmental/flaky all along**, and three data points (2 clean branches) isn't enough to rule that out given the low sample size — though 4/5 failures specifically on this branch, and 0/2 on two others, is a meaningful asymmetry.

The cheapest way to distinguish (2) from (1)/(3): merge current `main` into this branch and push. If `Build docs` goes green, that points squarely at (2) and this PR's own diff was never really at fault. I'll do that now unless you'd rather I hold off.

(Unrelated to the above: #403's `clang22-release_coverage` leg was cancelled at exactly the 240-minute job timeout on a cold cache — a from-main branch has no warm ccache to restart from. Not a real failure and doesn't affect the `Build docs` result, which had already completed by then.)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
Merged current `main` into this branch (`09f0956`) and pushed — no conflicts, clean merge, both existing diagnostics (`PYTHONFAULTHANDLER`, `dmesg`) survived intact. Watching the new run; this directly tests whether `Build docs` now passes here too.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by renefritze** [comment]
@renefritze Update on the `Build docs` `DeadKernelError` investigation: after merging current `main` into this branch (commit 09f0956), `Build docs (executes notebooks via myst-nb)` completed successfully: https://github.com/dune-gdt/dune-gdt/actions/runs/30225862695/job/89875933320

Summary of the investigation:
- Ruled out a trappable signal (`PYTHONFAULTHANDLER=1` showed nothing) and the OOM killer (`dmesg` was clean) as the cause of the kernel death.
- Two isolated diagnostic PRs both had `Build docs` pass cleanly: #402 (based on `main`, zero code from this PR) and #403 (based on `main`, containing only the `py::keep_alive` binding changes from this PR, without the estimator/exceptions changes).
- That pointed at either flakiness or something already fixed on `main` since this branch was created, rather than this PR's own diff being at fault.
- Merging `main` into this branch and re-running confirmed it: `Build docs` now passes here too.

`Build on Linux (clang22-debug)`, `Build wheels (3.13)`, and `SonarCloud` are all green. `clang22-release_coverage` is still `in_progress` (long-running coverage build); once it completes, the PR should be fully green. Auto-merge is already armed, so no manual merge is needed once CI finishes.

I'll leave #402 and #403 open for now in case they're useful reference, but they're no longer blocking — happy to close them if you'd rather not carry the extra CI load.

---
_Generated by [Claude Code](https://claude.ai/code)_
